Libraries/LibWeb/HTML/HTMLMediaElement.cpp

```diff
@@ -1395,6 +1356,67 @@ WebIDL::ExceptionOr<void> HTMLMediaElement::process_media_data(Function<void(Str
             return IterationDecision::Break;
         });
     }
+}
+
+// https://html.spec.whatwg.org/multipage/media.html#media-data-processing-steps-list
+WebIDL::ExceptionOr<void> HTMLMediaElement::setup_playback_manager(Function<void(String)> failure_callback)
+{
+    m_playback_manager = Media::PlaybackManager::create();
+
+    m_has_enabled_preferred_audio_track = false;
+    m_has_selected_preferred_video_track = false;
+
+    // -> If the media resource is found to have an audio track
+    // -> If the media resource is found to have a video track
+    m_playback_manager->on_track_added = [weak_self = GC::Weak(*this)](auto track_type, auto& track) {
+        if (!weak_self)
+            return;
+        if (track_type == Media::TrackType::Audio) {
+            weak_self->on_audio_track_added(track);
+        } else {
+            weak_self->on_video_track_added(track);
+        }
+    };
+
+    // -> Once enough of the media data has been fetched to determine the duration of the media resource, its dimensions, and other metadata
+    m_playback_manager->on_metadata_parsed = [weak_self = GC::Weak(*this)](auto duration) {
+        if (!weak_self)
+            return;
+        weak_self->on_metadata_parsed(duration);
+    };
+
+    // -> If the media data can be fetched but is found by inspection to be in an unsupported format, or can otherwise not be rendered at all
+    m_playback_manager->on_unsupported_format_error = [weak_self = GC::Weak(*this), failure_callback = move(failure_callback)](auto&& error) mutable {
+        if (!weak_self)
+            return;
+
+        // 1. The user agent should cancel the fetching process.
+        weak_self->m_fetch_controller->stop_fetch();
+
+        // 2. Abort this subalgorithm, returning to the resource selection algorithm.
+        failure_callback(MUST(String::from_utf8(error.description())));
+    };
+
+    m_playback_manager->add_media_source(m_media_data);
+
+    // NOTE: The spec is unclear on whether the following media resource track conditions should trigger multiple
+    //       times on one media resource, but it is implied to be possible by the start of the "Media elements"
+    //       section, where it says that a "media resource can have multiple audio and video tracks."
+    //       https://html.spec.whatwg.org/multipage/media.html#media-elements
+    //       Therefore, we enumerate all the available tracks into our VideoTrackList and AudioTrackList.
```

Comment:
This note could go above the calls to on_(audio|video)_track_added.

Reply:
Oh also, this is a non-spec comment, so if we've switched to using `NB:`, this should be changed.